### PR TITLE
Dawn Integration Part 2

### PR DESCRIPTION
**Pull request by eddie-c-davis**
## Description

This PR is the second portion of the integration with the Dawn compiler and includes implementations of the various Dawn backends for CPU and GPU.

### CODE REVIEW

File: src/gt4py/backend/pyext_builder.py [hunk 1]
```diff
@@ -51,6 +51,7 @@ def get_gt_pyext_build_opts(
         ],
         nvcc=[
             "-std=c++14",
+            "-arch=sm_35",
```
Comment (jdahm) [inline src/gt4py/backend/pyext_builder.py:54]: Is this required?

File: src/gt4py/config.py [hunk 2]
```diff
@@ -23,7 +23,7 @@
 
 # Default paths (taken from user's environment vars when possible)
 BOOST_ROOT: str = os.environ.get(
-    "BOOST_ROOT", os.environ.get("BOOST_HOME", os.path.abspath("/usr/local/include"))
+    "BOOST_ROOT", os.environ.get("BOOST_HOME", os.path.abspath("/usr/local"))
```
Comment (jdahm) [inline src/gt4py/config.py:26]: Is this how it is installed in Ubuntu? Could always be manually specified with `BOOST_HOME`.

File: src/gt4py/backend/pyext_builder.py (hunk 1 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/pyext_builder.py:54]: Yes, this is required for the `__ldg` intrinsic that is used by the `dawn:cuda` backend.

File: src/gt4py/config.py (hunk 2 above)
Comment (eddie-c-davis) [reply src/gt4py/config.py:26]: This change is because when the code executed, the path actually resolved to `/usr/local/include/include`. And yes, `BOOST_HOME` can be specified as an environmental variable.

File: src/gt4py/__init__.py [hunk 3]
```diff
@@ -37,6 +37,7 @@
 
 from .definitions import AccessKind, Boundary, DomainInfo, FieldInfo, ParameterInfo, CartesianSpace
 from .stencil_object import StencilObject
+from .test_builder import TestBuilder
```
Comment (jdahm) [inline src/gt4py/__init__.py:40]: `TestBuilder` seems like something that should not be imported by default.

File: src/gt4py/backend/pyext_builder.py (hunk 1 above)
Comment (jdahm) [reply src/gt4py/backend/pyext_builder.py:54]: Can we make this read the `CUDA_ARCH` env variable so it can be overridden with a newer architecture flag as possible?

File: src/gt4py/test_builder.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
+import sys
```
Comment (jdahm) [inline src/gt4py/test_builder.py:1]: Can the TestBuilder be a separate PR from the dawn integration part 2?

File: src/gt4py/config.py (hunk 2 above)
Comment (jdahm) [reply src/gt4py/config.py:26]: Good fix then!

Comment (eddie-c-davis) [reply src/gt4py/config.py:26]: Thanks!

File: src/gt4py/__init__.py (hunk 3 above)
Comment (eddie-c-davis) [reply src/gt4py/__init__.py:40]: Removed.

File: src/gt4py/test_builder.py (hunk 4 above)
Comment (eddie-c-davis) [reply src/gt4py/test_builder.py:1]: Removed.

File: src/gt4py/backend/unit_test.cpp.in [hunk 5]
```diff
@@ -0,0 +1,202 @@
+{#
```
Comment (jdahm) [inline src/gt4py/backend/unit_test.cpp.in:1]: Is this part of the TestBuilder?

File: src/gt4py/backend/pyext_builder.py (hunk 1 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/pyext_builder.py]: Implemented.

File: src/gt4py/backend/unit_test.cpp.in (hunk 5 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/unit_test.cpp.in]: Good catch, missed that one. Removed.

File: src/gt4py/backend/pyext_builder.py (hunk 1 above)
Comment (gronerl) [reply src/gt4py/backend/pyext_builder.py]: If not explicitly set in the environment, we should get it at runtime, e.g.: 
```
def get_cuda_compute_capability():
    try:
        import cupy as cp

        return cp.cuda.Device(0).compute_capability
    except:
        return None
```

Comment (gronerl) [reply src/gt4py/backend/pyext_builder.py]: Also, it would be nice if an error is raised if the device is too old for that backend.

Comment (egparedes) [reply src/gt4py/backend/pyext_builder.py]: I agree with Linus about detecting the compute capability at runtime if not explicitly provided. Actually, it might be better to raise an exception if the detection does not work to avoid compilations without setting a specific architecture. Something like:
```python
# config.py

"cuda_arch": os.environ.get("CUDA_ARCH", None)
```
and then in the `pyext_builder`:
```python
cuda_arch = gt_config.build_settings["cuda_arch"] or get_cuda_compute_capability()
if not cuda_arch:
    raise Exception(...) # Cuda architecture not set....
...

             "-arch={}".format(cuda_arch), 
```

File: src/gt4py/storage/storage.py [hunk 6]
```diff
@@ -399,6 +399,8 @@ def _construct(cls, backend, dtype, default_origin, shape, alignment, layout_map
 
     @property
     def data(self):
+        if cp is not None:
```
Comment (egparedes) [inline src/gt4py/storage/storage.py:402]: Why is this needed? I think this violates the convention that `storage.data` is returning the actual cupy/numpy buffer.

File: src/gt4py/backend/templates/dawn_computation.src.in [hunk 7]
```diff
@@ -41,17 +41,42 @@ using backend_t = DAWN_GT_BACKEND;
...
+    for (int i = 0; i < 3; ++i) {
+      if(min_origin[i] < halos_[i*2])
+          halos_[i*2] = halos_[i*2+1] = min_origin[i];
+      size[i] = halos_[i*2] + in_size[i] + halos_[i*2+1];
+    }
 
-gridtools::dawn::domain make_domain(const std::array<gt::uint_t, 3> &size) {
     gridtools::dawn::domain d{size};
-    d.set_halos(0, 0, 0, 0, 0, 0);
+    d.set_halos(halos_[0], halos_[1], halos_[2], halos_[3], halos_[4], halos_[5]);
```
Comment (egparedes) [inline src/gt4py/backend/templates/dawn_computation.src.in:79]: Is this working also working if the buffer is shifted when the user sets a different origin?

File: src/gt4py/backend/templates/dawn_computation.src.in [hunk 8]
```diff
@@ -93,7 +107,7 @@ data_store_t make_data_store(const BufferInfo& bi,
     double* ptr = static_cast<double*>(bi.ptr);
     for (int i = 0; i < 3; ++i) {
         strides[i] = bi.strides[i] / sizeof(double);
-        ptr += strides[i] * origin[i];
+        ptr += strides[i] * (origin[i] - halos_[i*2]);
```
Comment (egparedes) [inline src/gt4py/backend/templates/dawn_computation.src.in:110]: Same question as before: Do we really need to set halos in the `data_store` because the grid generated by Dawn requires them? In the gridtools backends we completely avoid them

**Comment by egparedes** [comment]
I still want to take a deeper look to the changes in `dawn_backends` but my main concerns with the PR have been raised already raised by @jdahm (and fixed), so it looks closer to be accepted.

One **important** thing that is missing is to verify that all the external people contributing to this PR have signed a contributor assignment agreement and to update the `AUTHORS.rst` file with their names.

File: src/gt4py/storage/storage.py (hunk 6 above)
Comment (eddie-c-davis) [reply src/gt4py/storage/storage.py:402]: This may be a problem on our side, the `fv3ser` test suite does not distinguish between numpy and cupy arrays. Could we add a convenience method to `storage`, such as `asnumpy`, that converts any storage to a numpy array?

File: src/gt4py/backend/templates/dawn_computation.src.in (hunk 8 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/templates/dawn_computation.src.in:110]: I will experiment with the same method as the gridtools backends.

File: src/gt4py/storage/storage.py (hunk 6 above)
Comment (gronerl) [reply src/gt4py/storage/storage.py:402]: you can always do
* `storage.view(np.ndarray)` since currently it's a subclass of np.ndarray
* `np.asarray(storage)`
note though that both of these options currently give you the cpu buffer as is, it does not trigger synchronization. 
you could do `storage.device_to_host()` before any of these options to make sure that you get an up-to-date cpu buffer. (it may later go out of sync again, so you shouldn't keep a reference longer than you are sure nothing is written in the storage. 

`device_to_host` actually also works for cpu storages, just does nothing there. 
we could in principle add that `np.asarray(storage)` triggers a synchronization.

Comment (eddie-c-davis) [reply src/gt4py/storage/storage.py:402]: Thank you @gronerl, `storage.device_to_host(force=True)` followed by `np.asarray(storage)` worked.

Comment (gronerl) [reply src/gt4py/storage/storage.py:402]: is the force necessary? this is potentially unsafe if the host buffer happens to be up to date already.

Comment (eddie-c-davis) [reply src/gt4py/storage/storage.py:402]: Yes, the code did not work without it, because `_is_device_modified` was not `True`. This may be an issue with the Dawn backends, I will look into it.

Comment (eddie-c-davis) [reply src/gt4py/storage/storage.py:402]: Found the issue, and removed the `force`, thank you for the assistance.

File: src/gt4py/backend/templates/dawn_computation.src.in (hunk 8 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/templates/dawn_computation.src.in:110]: Setting the compute domain halos equal to the origin and removing the pointer increments allows the halos to be avoided as in the gridtools backends. Thank you for the suggestion.

File: AUTHORS.rst [hunk 9]
```diff
@@ -6,6 +6,7 @@ Authors
 
 * Benedicic, Lucas. ETH Zurich - CSCS
 * Cruz, Felipe. ETH Zurich - CSCS
+* Davis, Eddie. Vulcan, Inc.
```
Comment (jdahm) [inline AUTHORS.rst:9]: I was told that "Vulcan, Inc." should not have a comma in it.
```suggestion
* Davis, Eddie. Vulcan Inc.
```

**Comment by egparedes** [comment]
daw4py dependency in `setup.cfg` should be pinned to a specific version by adding `@tag` in the url before `#subdirectory=dawn`. For example: `dawn4py@git+https://github.com/MeteoSwiss-APN/dawn.git@0.0.2#subdirectory=dawn`

File: src/gt4py/backend/dawn_backends.py [hunk 10]
```diff
@@ -15,27 +15,42 @@
 # SPDX-License-Identifier: GPL-3.0-or-later
 
 import abc
+import enum
 import inspect
 import numbers
 import os
 import re
 import types
+import copy
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:24]: please sort the import alphabetically and add `from typing ...` from below as the last line of this block, since `typing` is part of the standard library

File: src/gt4py/backend/dawn_backends.py [hunk 11]
```diff
@@ -15,27 +15,42 @@
...
 
 import jinja2
+import numpy as np
+
+from typing import List, Dict, Type
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:29]: move to first block of imports

File: src/gt4py/backend/dawn_backends.py [hunk 12]
```diff
@@ -86,6 +103,20 @@ def _make_global_variables(self, parameters: list, externals: dict):
 
         return global_variables
 
+    def visit_Operator(self, op: enum.Enum):
+        if op in self.OP_TO_CPP:
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:107]: A  shorter way to express this: `self.OP_TO_CPP.get(op, op.symbol)`

File: src/gt4py/backend/dawn_backends.py [hunk 13]
```diff
@@ -107,19 +138,38 @@ def visit_VarRef(self, node: gt_ir.VarRef, **kwargs):
...
-        op = node.op.python_symbol
-        return sir_utils.make_binary_operator(left, op, right)
+        if node.op.python_symbol == "**":
+            sir = self.visit_ExpOpExpr(left, right)
+        else:
+            op = self.visit_Operator(node.op)
+            sir = sir_utils.make_binary_operator(left, op, right)
+        return sir
+
+    def visit_ExpOpExpr(self, left, right):
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:159]: This is a helper method not really a visit method for `ExpOpExpr` nodes, so it should be renamed to make this explicit (like `make_pow_expr` or something similar)

File: src/gt4py/backend/dawn_backends.py [hunk 14]
```diff
@@ -107,19 +138,38 @@ def visit_VarRef(self, node: gt_ir.VarRef, **kwargs):
...
+            return sir_utils.make_literal_access_expr("1", type=SIR.BuiltinType.Integer)
+        if exponent == "1":
+            return sir_utils.make_unary_operator("+", left)
+        if exponent == "2":
+            return sir_utils.make_binary_operator(left, "*", left)
+        elif exponent == "3":
+            return sir_utils.make_binary_operator(
+                left, "*", sir_utils.make_binary_operator(left, "*", left)
+            )
+        else:
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:171]: Should the frontend take care of this optimization? I think this kind of analysis and substitutions should be done by the toolchain.

File: src/gt4py/backend/dawn_backends.py [hunk 15]
```diff
@@ -196,24 +251,113 @@ def visit_StencilDefinition(self, node: gt_ir.StencilDefinition, **kwargs):
...
+# Load or generate a Dawn Computation object for the current domain size
+{empty_checks}
+{none_checks}
+pyext_module.run_computation(list(_domain_), {run_args}, exec_info)
+""".format(
+            empty_checks="\n".join(empty_checks),
+            none_checks="\n".join(none_checks),
+            run_args=", ".join(args),
+        )
+        sources.extend(source.splitlines())
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:319]: I guess this code is creating default values for the case where optional parameters/fields have not been provided by the user, right? Could you add some comment here to make it clear what this code does and why ?

File: src/gt4py/backend/dawn_backends.py [hunk 16]
```diff
@@ -238,41 +383,43 @@ def generate(
...
             raise RuntimeError("Missing GridTools sources.")
-
-        module_kwargs = {}
         pyext_module_name, pyext_file_path = cls.generate_extension(
-            stencil_id, definition_ir, options, module_kwargs=module_kwargs
+            stencil_id, definition_ir, options, **kwargs
         )
 
-        # Generate and return the Python wrapper class
+        if not options._impl_opts.get("disable-code-generation", False):
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:399]: I know the hierarchy of classes and methods is convoluted and it is not at all well documented, but the idea here is that you external backends that have their own analysis toolchain can pass a `module_info` dictionary with the information required to build the python module instead of running the internal analysis pipeline to generate the `implementation_ir`. Look at the `__call__` method of the `BaseModuleGenerator` in `base.py` for more info of how this `module_info` will be used.

File: src/gt4py/backend/dawn_backends.py [hunk 17]
```diff
@@ -86,6 +103,20 @@ def _make_global_variables(self, parameters: list, externals: dict):
 
         return global_variables
 
+    def visit_Operator(self, op: enum.Enum):
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:106]: This is not a real `visit_` method to dispatch to Operator node, so please rename it to something like `make_operator` to make clear that this is a helper method.

File: src/gt4py/backend/dawn_backends.py [hunk 18]
```diff
@@ -139,6 +189,11 @@ def visit_Assign(self, node: gt_ir.Assign, **kwargs):
         stmt = sir_utils.make_assignment_stmt(left, right, "=")
         return stmt
 
+    def visit_AugAssign(self, node: gt_ir.AugAssign):
+        bin_op = gt_ir.BinOpExpr(lhs=node.target, op=node.op, rhs=node.value)
+        assign = gt_ir.Assign(target=node.target, value=bin_op)
+        return self.visit_Assign(assign)
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:195]: This should also work calling the general dispatcher method `self.visit` instead of calling directly the handler of `Assign` nodes, and it is more robust.

File: src/gt4py/backend/dawn_backends.py (hunk 16 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py:399]: I will work on this.

File: src/gt4py/backend/dawn_backends.py (hunk 15 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py:319]: Comments have been updated.

File: src/gt4py/backend/dawn_backends.py (hunk 14 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py:171]: Would this be a new class in `gtscript_frontend`, e.g., `PowerExprOptimizer`?

File: src/gt4py/backend/dawn_backends.py (hunk 13 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py:159]: Changed to a helper method.

File: src/gt4py/backend/dawn_backends.py (hunk 12 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py:107]: Updated as suggested.

File: src/gt4py/backend/dawn_backends.py (hunk 11 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py:29]: Imports have been reorganized.

File: src/gt4py/backend/dawn_backends.py (hunk 14 above)
Comment (egparedes) [reply src/gt4py/backend/dawn_backends.py]: My view on this is that this optimization does not belong to the frontend. I think the frontend should create an IR according to the DSL code, without introducing optimizations which can be applied later by the analysis toolchain.

Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py]: Thank you. We should wait for the math functions to be implemented before adding this optimization to the toolchain. I will remove it for now.

File: src/gt4py/backend/dawn_backends.py (hunk 16 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py:402]: The code has been refactored to generate a `module_info` dictionary rather than generating an `implementation_ir` using the `gt4py` pipeline.

File: src/gt4py/backend/base.py [hunk 19]
```diff
@@ -670,12 +670,13 @@ def __call__(
         stencil_id: gt_definitions.StencilID,
         definition_ir: gt_ir.StencilDefinition,
         options: gt_definitions.BuildOptions,
+        module_info: Optional[Dict[str, Any]] = None,
```
Comment (egparedes) [inline src/gt4py/backend/base.py:673]: Good catch!

File: src/gt4py/backend/dawn_backends.py [hunk 20]
```diff
@@ -58,8 +76,23 @@ def visit_FieldDecl(self, node: gt_ir.FieldDecl, **kwargs):
             )
         )
 
+    def visit_FieldRef(self, node: gt_ir.FieldRef, **kwargs):
+        field_name = node.name
+        if self.in_write:
```
Comment (egparedes) [inline src/gt4py/backend/dawn_backends.py:81]: The generic node visitors always forward the received `**kwargs` and I think is good practice for custom `visit_` method handlers to do it too. I find convenient to pass an optional keyword argument with this information to the inner `visit_` method from the outer one, for this kind of local (in/out) state tracking. For example in this case you can avoid the `self.in_write` var by just doing:

```
def visit_FieldRef(self, node: gt_ir.FieldRef, *, in_write: bool = False, **kwargs):
    ...
    if in_write:
     ....
```
and then in line 88:
```left = self.visit(node.target, in_write=True, **kwargs)```

This should work perfectly if you always pass the `**kwargs` to the dispatcher `self.visit` method: `self.visit(..., **kwargs)`

Don't take this as a request to change the code but just as a comment for your information. Feel free to keep it as it is right now if it makes more sense to you.

**Comment by egparedes** [comment]
I added a comment to your last commit but it is mostly informative, just take a look. The current state of the PR is good for me so if no one else has complaints, I will merge the PR in 24 hours.

File: src/gt4py/backend/dawn_backends.py (hunk 20 above)
Comment (eddie-c-davis) [reply src/gt4py/backend/dawn_backends.py:81]: Thank you Enrique, that is an excellent suggestion.

**Comment by egparedes** [comment]
Ooops! I really should've caught this one! 😥
> ```
> src/gt4py/backend/dawn_backends.py:376:76: E711 comparison to None should be 'if cond is None:'
> ```
